### ISSUE TITLE
kube-apiserver 1.22.4 panics with fatal error: concurrent map iteration and map write

### ISSUE THREAD

**Issue by JohnRusk** [kubernetes/kubernetes#107463]
### What happened?

I'm doing some load testing with K8s 1.22.4.  In particular, restarting the API server with thousands of nodes.  The kubelets to re-list to restart their watches, and sometimes the API server panics and crashes.  I captured the error and it was:

```
      fatal error: concurrent map iteration and map write
running]:
runtime.throw(0x4f097c2, 0x26)
/usr/local/go/src/runtime/panic.go:1117 +0x72 fp=0xc20eb46f18 sp=0xc20eb46ee8 pc=0x4388d2
runtime.mapiternext(0xc20eb47008)
/usr/local/go/src/runtime/map.go:858 +0x54c fp=0xc20eb46f98 sp=0xc20eb46f18 pc=0x410fcc
runtime.mapiterinit(0x4cace40, 0xc86b1018f0, 0xc20eb47008)
/usr/local/go/src/runtime/map.go:848 +0x1c5 fp=0xc20eb46fb8 sp=0xc20eb46f98 pc=0x410985
k8s.io/kubernetes/vendor/golang.org/x/net/http2.cloneHeader(0xc86b1018f0, 0xc8ba301800)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/server.go:2658 +0xba fp=0xc20eb47078 sp=0xc20eb46fb8 pc=0x9bc85a
k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*responseWriterState).writeHeader(0xc8ee912e80, 0x1f8)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/server.go:2651 +0x6e fp=0xc20eb470d0 sp=0xc20eb47078 pc=0x9bc6ce
    
```

### What did you expect to happen?

API Server does not panic

### How can we reproduce it (as minimally and precisely as possible)?

You might be able to reproduce it by running the API server with relatively high load in a relatively resource constrained environment: maybe something like 2000 nodes, with API server on a 16 core host machine.  (Maybe you could repro it with a bigger server or fewer nodes too, but it seems to repro most easily when the server is under load).

### Anything else we need to know?

My test server is running as a pod in another K8s cluster, although I'm not aware of any reason what would contribute to the problem.

### Kubernetes version

<details>

1.22.4 

</details>

### Cloud provider

<details>

Azure

</details>

### OS version

<details>

n/a

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by MadhavJivrajani** [comment]
@JohnRusk 👋🏼 
Is it possible to provide more of the stack trace? I see that the panic is caused by concurrent invocation of iteration in `cloneHeader`, called through `writeHeader`. I'd like to try and narrow down the code surface area of api-server that might be causing this.

**Comment by fedebongio** [comment]
/cc @wojtek-t 
/sig scalability
@JohnRusk would be great to have stack trace +1 
/triage accepted

**Comment by JohnRusk** [comment]
@MadhavJivrajani @fedebongio here is the full callstack

```

running]:
runtime.throw(0x4f097c2, 0x26)
	/usr/local/go/src/runtime/panic.go:1117 +0x72 fp=0xc20eb46f18 sp=0xc20eb46ee8 pc=0x4388d2
runtime.mapiternext(0xc20eb47008)
	/usr/local/go/src/runtime/map.go:858 +0x54c fp=0xc20eb46f98 sp=0xc20eb46f18 pc=0x410fcc
runtime.mapiterinit(0x4cace40, 0xc86b1018f0, 0xc20eb47008)
	/usr/local/go/src/runtime/map.go:848 +0x1c5 fp=0xc20eb46fb8 sp=0xc20eb46f98 pc=0x410985
k8s.io/kubernetes/vendor/golang.org/x/net/http2.cloneHeader(0xc86b1018f0, 0xc8ba301800)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/server.go:2658 +0xba fp=0xc20eb47078 sp=0xc20eb46fb8 pc=0x9bc85a
k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*responseWriterState).writeHeader(0xc8ee912e80, 0x1f8)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/server.go:2651 +0x6e fp=0xc20eb470d0 sp=0xc20eb47078 pc=0x9bc6ce
k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*responseWriter).WriteHeader(0xc8cb4600b8, 0x1f8)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/server.go:2642 +0x3d fp=0xc20eb470f0 sp=0xc20eb470d0 pc=0x9bc61d
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc8d6df4aa0, 0x1f8)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:269 +0x5c fp=0xc20eb47110 sp=0xc20eb470f0 pc=0x19c8b3c
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.withRequestDeadline.func1(0x56af5b8, 0xc8d6df4aa0, 0xc8dafd5600)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/request_deadline.go:101 +0x3b7 fp=0xc20eb47400 sp=0xc20eb472a0 pc=0x19da077
net/http.HandlerFunc.ServeHTTP(0xc03c2af730, 0x56af5b8, 0xc8d6df4aa0, 0xc8dafd5600)
	/usr/local/go/src/net/http/server.go:2049 +0x44 fp=0xc20eb47428 sp=0xc20eb47400 pc=0x887264
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithWaitGroup.func1(0x56af5b8, 0xc8d6df4aa0, 0xc8dafd5600)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/waitgroup.go:59 +0x137 fp=0xc20eb476a0 sp=0xc20eb47428 pc=0x1bc8ef7
net/http.HandlerFunc.ServeHTTP(0xc03c2f9cb0, 0x56af5b8, 0xc8d6df4aa0, 0xc8dafd5600)
	/usr/local/go/src/net/http/server.go:2049 +0x44 fp=0xc20eb476c8 sp=0xc20eb476a0 pc=0x887264
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuditAnnotations.func1(0x56af5b8, 0xc8d6df4aa0, 0xc8dafd5400)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/audit_annotations.go:37 +0x142 fp=0xc20eb47730 sp=0xc20eb476c8 pc=0x19d5142
net/http.HandlerFunc.ServeHTTP(0xc03c0ff800, 0x56af5b8, 0xc8d6df4aa0, 0xc8dafd5400)
	/usr/local/go/src/net/http/server.go:2049 +0x44 fp=0xc20eb47758 sp=0xc20eb47730 pc=0x887264
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithWarningRecorder.func1(0x56af5b8, 0xc8d6df4aa0, 0xc8dafd5200)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/warning.go:35 +0x1a8 fp=0xc20eb477e0 sp=0xc20eb47758 pc=0x19dc168
net/http.HandlerFunc.ServeHTTP(0xc03c0ff818, 0x56af5b8, 0xc8d6df4aa0, 0xc8dafd5200)
	/usr/local/go/src/net/http/server.go:2049 +0x44 fp=0xc20eb47808 sp=0xc20eb477e0 pc=0x887264
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithCacheControl.func1(0x56af5b8, 0xc8d6df4aa0, 0xc8dafd5200)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/cachecontrol.go:31 +0xa8 fp=0xc20eb47878 sp=0xc20eb47808 pc=0x19d7268
net/http.HandlerFunc.ServeHTTP(0xc03c0ff830, 0x56af5b8, 0xc8d6df4aa0, 0xc8dafd5200)
	/usr/local/go/src/net/http/server.go:2049 +0x44 fp=0xc20eb478a0 sp=0xc20eb47878 pc=0x887264
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.WithLogging.func1(0x56a4ab8, 0xc8cb4600b8, 0xc8dafd5100)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:108 +0x37e fp=0xc20eb479c8 sp=0xc20eb478a0 pc=0x19c917e
net/http.HandlerFunc.ServeHTTP(0xc03c30a300, 0x56a4ab8, 0xc8cb4600b8, 0xc8dafd5100)
	/usr/local/go/src/net/http/server.go:2049 +0x44 fp=0xc20eb479f0 sp=0xc20eb479c8 pc=0x887264
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithRequestInfo.func1(0x56a4ab8, 0xc8cb4600b8, 0xc8dafd4f00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x287 fp=0xc20eb47a98 sp=0xc20eb479f0 pc=0x19db3e7
net/http.HandlerFunc.ServeHTTP(0xc03c2f9ce0, 0x56a4ab8, 0xc8cb4600b8, 0xc8dafd4f00)
	/usr/local/go/src/net/http/server.go:2049 +0x44 fp=0xc20eb47ac0 sp=0xc20eb47a98 pc=0x887264
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.withRequestReceivedTimestampWithClock.func1(0x56a4ab8, 0xc8cb4600b8, 0xc8dafd4e00)
net/http.HandlerFunc.ServeHTTP(0xc03c2f9d10, 0x56a4ab8, 0xc8cb4600b8, 0xc8dafd4e00)
	/usr/local/go/src/net/http/server.go:2049 +0x44 fp=0xc20eb47b78 sp=0xc20eb47b50 pc=0x887264
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.withPanicRecovery.func1(0x56a4ab8, 0xc8cb4600b8, 0xc8dafd4e00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/wrap.go:74 +0xe6 fp=0xc20eb47c00 sp=0xc20eb47b78 pc=0x1bc9e06
net/http.HandlerFunc.ServeHTTP(0xc03c30a320, 0x56a4ab8, 0xc8cb4600b8, 0xc8dafd4e00)
	/usr/local/go/src/net/http/server.go:2049 +0x44 fp=0xc20eb47c28 sp=0xc20eb47c00 pc=0x887264
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.withAuditID.func1(0x56a4ab8, 0xc8cb4600b8, 0xc8db413e00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go:66 +0x20a fp=0xc20eb47cf8 sp=0xc20eb47c28 pc=0x19dc62a
net/http.HandlerFunc.ServeHTTP(0xc03c30a340, 0x56a4ab8, 0xc8cb4600b8, 0xc8db413e00)
	/usr/local/go/src/net/http/server.go:2049 +0x44 fp=0xc20eb47d20 sp=0xc20eb47cf8 pc=0x887264
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.(*APIServerHandler).ServeHTTP(0xc03c2f9d40, 0x56a4ab8, 0xc8cb4600b8, 0xc8db413e00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:189 +0x51 fp=0xc20eb47d50 sp=0xc20eb47d20 pc=0x1cc21b1
net/http.serverHandler.ServeHTTP(0xc038655180, 0x56a4ab8, 0xc8cb4600b8, 0xc8db413e00)
	/usr/local/go/src/net/http/server.go:2867 +0xa3 fp=0xc20eb47d80 sp=0xc20eb47d50 pc=0x88a823
net/http.initALPNRequest.ServeHTTP(0x56b45f8, 0xc86a20d290, 0xc90d6d1500, 0xc038655180, 0x56a4ab8, 0xc8cb4600b8, 0xc8db413e00)
	/usr/local/go/src/net/http/server.go:3439 +0x8d fp=0xc20eb47ef8 sp=0xc20eb47d80 pc=0x88d44d
net/http.(*initALPNRequest).ServeHTTP(0xc88e704a20, 0x56a4ab8, 0xc8cb4600b8, 0xc8db413e00)
	<autogenerated>:1 +0x75 fp=0xc20eb47f40 sp=0xc20eb47ef8 pc=0x8b7975
net/http.Handler.ServeHTTP-fm(0x56a4ab8, 0xc8cb4600b8, 0xc8db413e00)
	/usr/local/go/src/net/http/server.go:87 +0x56 fp=0xc20eb47f70 sp=0xc20eb47f40 pc=0x8b0976
k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*serverConn).runHandler(0xc8673f5b00, 0xc8cb4600b8, 0xc8db413e00, 0xc65b313c20)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/server.go:2162 +0x8b fp=0xc20eb47fc0 sp=0xc20eb47f70 pc=0x9ba4eb
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc20eb47fc8 sp=0xc20eb47fc0 pc=0x472701
created by k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*serverConn).processHeaders
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/server.go:1892 +0x505

```

**Comment by JohnRusk** [comment]
BTW, the last things logged before the panic were
```
E0110 03:21:17.198751       1 repair.go:118] unable to refresh the service IP block: Get "https://localhost:443/api/v1/services": stream error: stream ID 997; INTERNAL_ERROR
E0110 03:21:17.200033       1 repair.go:75] unable to refresh the port block: Get "https://localhost:443/api/v1/services": stream error: stream ID 999; INTERNAL_ERROR

```

**Comment by aojea** [comment]
> BTW, the last things logged before the panic were
> 
> ```
> E0110 03:21:17.198751       1 repair.go:118] unable to refresh the service IP block: Get "https://localhost:443/api/v1/services": stream error: stream ID 997; INTERNAL_ERROR
> E0110 03:21:17.200033       1 repair.go:75] unable to refresh the port block: Get "https://localhost:443/api/v1/services": stream error: stream ID 999; INTERNAL_ERROR
> ```

Those messages are trying to get the bitmaps from etcd with the IP and nodeport allocation information.

The trace seems that is a timeout.

I would check that etcd is responsive, it may happen that is overwhelmed and not being able to communicate correctly with the apiservers

**Comment by JohnRusk** [comment]
aojea I assume that any concurrent map iteration and write would indicate a bug. That's never supposed to happen, right?  That's why it panics when it does, isn't it.  Even if it does in fact happen if etcd is busy, it still shouldn't happen.

**Comment by aojea** [comment]
> aojea I assume that any concurrent map iteration and write would indicate a bug. That's never supposed to happen, right? That's why it panics when it does, isn't it. Even if it does in fact happen if etcd is busy, it still shouldn't happen.

upps, yeah, I was reasoning out loud about those messages ... trying to understand if we can have a reproducer in an unit or integration test ... then things are easier to fix

**Comment by aojea** [comment]
I knew I saw something similar recently, I think that this may be related

https://github.com/kubernetes/kubernetes/pull/105734

**Comment by wojtek-t** [comment]
There was one more in the same category in last months:
https://github.com/kubernetes/kubernetes/issues/105759

Both are addressed at head.

However, from the stacktrace:
```
k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*responseWriter).WriteHeader(0xc8cb4600b8, 0x1f8)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/server.go:2642 +0x3d fp=0xc20eb470f0 sp=0xc20eb470d0 pc=0x9bc61d
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc8d6df4aa0, 0x1f8)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:269 +0x5c fp=0xc20eb47110 sp=0xc20eb470f0 pc=0x19c8b3c
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.withRequestDeadline.func1(0x56af5b8, 0xc8d6df4aa0, 0xc8dafd5600)
```

Looking into 1.18 release (as well as 1.17 and 1.19) I don't see httplog.go file to even have 269 lines:
https://github.com/kubernetes/kubernetes/blob/release-1.18/staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go

Are you running a modified fork?

@liggitt - FYI

**Comment by wojtek-t** [comment]
Never mind - I don't know where did I get 1.18 from. In 1.22 it all makes sense.

I'm just wondering if the above two fixed ( #105734 and #105759 ) actually addresses this.

**Comment by JohnRusk** [comment]
Thanks @wojtek-t. I can't test on head or 1.23 right now, but it's good to know that the issue may be fixed there.

**Comment by liggitt** [comment]
the cloneHeader panic matches what was fixed in https://github.com/kubernetes/kubernetes/pull/107452 and released in v1.22.6
